### DISCUSSION TITLE
Issues with a custom tap and the CI jobs

### DISCUSSION THREAD

**Discussion by csparker247** [Homebrew/discussions#4284] [Tap maintenance and Homebrew development]
### Output of `brew config`
<details>
<summary>My issues are with the CI actions, not my personal config, so collapsing</summary>

```shell
HOMEBREW_VERSION: 4.0.4-86-g7c15dce
ORIGIN: https://github.com/Homebrew/brew
HEAD: 7c15dce285900c85476d445dcbc01418e61be37c
Last commit: 6 hours ago
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 703d8de960be1944e0b24fabd83d9884b78f1939
Core tap last commit: 3 hours ago
Core tap branch: master
Core tap JSON: 28 Feb 21:18 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: /private/tmp/com.apple.launchd.pOzhECoDeo/org.xquartz:0
HOMEBREW_MAKE_JOBS: 10
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: 10-core 64-bit arm_firestorm_icestorm
Clang: 14.0.0 build 1400
Git: 2.39.2 => /opt/homebrew/bin/git
Curl: 7.86.0 => /usr/bin/curl
macOS: 13.2.1-arm64
CLT: 14.2.0.0.1.1668646533
Xcode: N/A
Rosetta 2: false
```
</details>

### Output of `brew doctor`
<details>
<summary>Same as above</summary>

```shell
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Some installed formulae are deprecated or disabled.
You should find replacements for the following formulae:
  ufraw

Warning: Unbrewed dylibs were found in /usr/local/lib.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected dylibs:
  /usr/local/lib/libCppUnit_clang90_v3_0.dylib
  /usr/local/lib/libGCBaseTest_clang90_v3_0.dylib
  /usr/local/lib/libGCBase_clang90_v3_0.dylib
  /usr/local/lib/libGenApi_clang90_v3_0.dylib
  /usr/local/lib/libLog_clang90_v3_0.dylib
  /usr/local/lib/libMathParser_clang90_v3_0.dylib
  /usr/local/lib/libNodeMapData_clang90_v3_0.dylib
  /usr/local/lib/libResUsageStat_clang90_v3_0.dylib
  /usr/local/lib/libXmlParser_clang90_v3_0.dylib
  /usr/local/lib/liblog4cpp_clang90_v3_0.dylib
  /usr/local/lib/libomp.dylib
  /usr/local/lib/libusb-1.0.0.dylib
  /usr/local/lib/libusb-1.0.dylib

Warning: Unbrewed header files were found in /usr/local/include.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected header files:
  /usr/local/include/spinnaker/[a whole bunch of files related to FLIRs Spinnaker API]
  ...

Warning: You have unlinked kegs in your Cellar.
Leaving kegs unlinked can lead to build-trouble and cause formulae that depend on
those kegs to fail to run properly once built. Run `brew link` on these:
  pkg-config
  xorgproto
  xvid
  opus
  boost
  openjpeg
  xmlto
  zimg
  speex
  shared-mime-info
  zeromq
  openexr
  spdlog
  opencv
  snappy
  opencore-amr
  pcre
  x264
  x265
  itk

Warning: Some taps are not on the default git origin branch and may not receive
updates. If this is a surprise to you, check out the default branch with:
  git -C $(brew --repo educelab/code) checkout main
```

</details>

### Description of issue

I've been trying to setup a [custom tap](https://github.com/educelab/homebrew-code) for formula and taps, but have been having some trouble understanding the expected workflow and CI jobs. My first [formula](https://github.com/educelab/homebrew-code/pull/2) is a header-only library that is configured and installed with CMake but only optionally uses CMake after install. I've been generally following the [nlohmann-json formula](https://github.com/Homebrew/homebrew-core/blob/master/Formula/nlohmann-json.rb) since it's setup similarly. 

My first issue is with dependencies. If I setup my formula like I expect, the `ubuntu-22.04` test bot [fails](https://github.com/educelab/homebrew-code/actions/runs/4296650095/jobs/7488629186) with `Error: No available formula with the name "cmake"`:
```ruby
  depends_on "cmake" => :build
  depends_on "eigen" => [:build, :test]
```

If I change this so that CMake is required for both build and test, everything completes successfully, but I'm honestly not sure why?
```ruby
  depends_on "cmake" => [:build, :test]
  depends_on "eigen" => [:build, :test]
```

Second, once my formula appears to be working and passing tests correctly, I add the `pr-pull` tag. This action runs for a bit and then [fails](https://github.com/educelab/homebrew-code/actions/runs/4297432940/jobs/7490399687) with:
```
Error: 4 problems in 1 formula detected
Error: Can't find dependency '"cmake"'.
Error: Can't find dependency '"eigen"'.
Error: Can't find dependency '"cmake"'.
Error: Can't find dependency '"eigen"'.
Error: Failure while executing; `/home/linuxbrew/.linuxbrew/bin/brew audit --skip-style --debug educelab/code/openabf` exited with 1.
```

Finally, the `nlohmann-json` package has a platform independent bottle because it's just a header-only library. Can I do that that, too? Is this just done by manually running `brew bottle`?

**Comment by Bo98** [comment]
I'm looking into the `brew test` issue but `brew audit` should have already been fixed - I'll have a look. A (slow) workaround is to tap Homebrew/core.

> Finally, the `nlohmann-json` package has a platform independent bottle because it's just a header-only library. Can I do that that, too? Is this just done by manually running `brew bottle`?

If you have two or more bottles and all of them have the same sha256 then it should auto-combine them when running `pr-pull`.

**Reply by csparker247** [reply]
Excellent! I can confirm that adding `brew tap Homebrew/core` to the Pull bottles step fixed the publish action, though yes, haha, it is slow.

I'll have more formulae to add over the coming days, so I'll keep an eye out for any updates. Thanks for the help and for looking into it!
